File: packages/ios-client/Sources/ios-client/ClientBase.swift
import Foundation
import Logging
import WebRTC
import os

extension RTCPeerConnection {
    // currently `Membrane RTC Engine` can't handle track of diretion `sendRecv` therefore
    // we need to change all `sendRecv` to `sendOnly`.
    func enforceSendOnlyDirection() {
        for transceiver in transceivers {
            if transceiver.direction == .sendRecv {
                transceiver.setDirection(.sendOnly, error: nil)
            }
        }
    }
}

extension RTCRtpEncodingParameters {
    static func create(rid: String, active: Bool, scaleResolutionDownBy: NSNumber) -> RTCRtpEncodingParameters {
        let encoding = RTCRtpEncodingParameters()
        encoding.rid = rid
        encoding.isActive = active
        encoding.scaleResolutionDownBy = scaleResolutionDownBy
        return encoding
    }

    static func create(active: Bool) -> RTCRtpEncodingParameters {
        let encoding = RTCRtpEncodingParameters()
        encoding.isActive = active
        return encoding
    }
}

public protocol PlayerListener: AnyObject {
    func onTrackAdded(track: RTCVideoTrack)
    func onTrackRemoved(track: RTCVideoTrack)
}

protocol Player {
    var delegate: PlayerListener? { get set }

    func connect() async throws
    func disconnect()
}

public class ClientBase: NSObject, RTCPeerConnectionDelegate {
    private let stunServerUrl: String
    var patchEndpoint: String?
    var peerConnection: RTCPeerConnection?
    var iceCandidates: [RTCIceCandidate] = []
    var isConnectionSetUp: Bool = false
    var connectOptions: ClientConnectOptions?

    public var peerConnectionState: RTCPeerConnectionState? {
        return peerConnection?.connectionState
    }

    @Published public var videoTrack: RTCVideoTrack? {
        willSet {
            if let track = videoTrack {
                delegate?.onTrackRemoved(track: track)
            }
        }
        didSet {
            if let track = videoTrack {
                delegate?.onTrackAdded(track: track)
            }
        }
    }

    public var audioTrack: RTCAudioTrack?

    public weak var delegate: PlayerListener?
    public var onConnectionStateChanged: ((RTCPeerConnectionState) -> Void)?

    let logger = Logger(label: "com.swmansion.whipwhepclient")

    static var peerConnectionFactory = RTCPeerConnectionFactory(
        encoderFactory: RTCDefaultVideoEncoderFactory(),
        decoderFactory: RTCDefaultVideoDecoderFactory()
    )

    public init(stunServerUrl: String?) {
        self.stunServerUrl = stunServerUrl ?? "stun:stun.l.google.com:19302"
    }

    func setUpPeerConnection() {
        let stunServerUrl = stunServerUrl
        let stunServer = RTCIceServer(urlStrings: [stunServerUrl])
        let iceServers = [stunServer]

        let config = RTCConfiguration()
        config.iceServers = iceServers
        config.sdpSemantics = .unifiedPlan
        config.continualGatheringPolicy = .gatherContinually
        config.candidateNetworkPolicy = .all
        config.tcpCandidatePolicy = .disabled

        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        peerConnection = ClientBase.peerConnectionFactory.peerConnection(
            with: config,
            constraints: constraints,
            delegate: self
        )

        if peerConnection! == nil {
            print("Failed to establish RTCPeerConnection. Check initial configuration")
        }

        isConnectionSetUp = true
    }

    public func connect(_ connectOptions: ClientConnectOptions) async throws {
        self.connectOptions = connectOptions
    }

    /**
     Sends an SDP offer to the WHIP/WHEP server.
    
     - Parameter sdpOffer: The offer to send to the server.
    
     - Throws: `AttributeNotFoundError.ResponseNotFound` if there is no response to the offer or
      `AttributeNotFoundError.LocationNotFound` if the response does not contain the location parameter or
      `SessionNetworkError.ConnectionError` if the  connection could not be established or the response code is incorrect,
       for example due to server being down, wrong server URL or token.
    
     - Returns: A SDP response.
     */
    func send(sdpOffer: String) async throws -> String {
        guard let connectOptions else {
            throw SessionNetworkError.ConnectionError(
                description:
                    "Cannot send the SDP Offer. Connection not setup. Remember to call connect first.")
        }

        var request = URLRequest(url: connectOptions.serverUrl)
        request.httpMethod = "POST"
        request.httpBody = sdpOffer.data(using: .utf8)
        request.addValue("application/sdp", forHTTPHeaderField: "Accept")
        request.addValue("application/sdp", forHTTPHeaderField: "Content-Type")
        if let token = connectOptions.authToken {
            request.addValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        var data: Data?
        var response: URLResponse?
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw SessionNetworkError.ConnectionError(
                description:
                    "Network error. Check if the server is up and running and the token and the server url is correct.")
        }
        guard let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 201
        else {
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw SessionNetworkError.ConnectionError(
                description:
                    "Network error. Check if the server is up and running and the token and the server url is correct.")
        }

        let responseString = String(data: data!, encoding: .utf8)
        if let foundLocation = httpResponse.allHeaderFields["Location"] as? String {
            patchEndpoint = foundLocation
            print("Location: \(foundLocation)")
        } else {
            throw AttributeNotFoundError.LocationNotFound(
                description: "Location attribute not found. Check if the SDP answer contains location parameter.")
        }
        guard let response = responseString else {
            throw AttributeNotFoundError.ResponseNotFound(
                description: "Response to SDP offer not found. Check if the network request was successful.")
        }
        return response
    }

    /**
     Sends an ICE candidate to WHIP/WHEP server in order to provide a streaming device.
    
     - Parameter candidate: Represents a single ICE candidate.
    
     - Throws: `AttributeNotFoundError.PatchEndpointNotFound` if the patch endpoint has not been properly set up,
       `AttributeNotFoundError.UFragNotFound` if the SDP of the candidate does not contain the ufrag,
       `SessionNetworkError.CandidateSendingError` if the candidate could not be sent and
       `NSError` for when the candidate data dictionary could not be serialized to JSON.
     */
    func sendCandidate(candidate: RTCIceCandidate) async throws {
        guard let connectOptions else {
            throw SessionNetworkError.ConnectionError(
                description:
                    "Cannot send ICE Candidate. Connection not setup. Remember to call connect first.")
        }

        guard patchEndpoint != nil else {
            throw AttributeNotFoundError.PatchEndpointNotFound(
                description: "Patch endpoint not found. Make sure the SDP answer is correct.")
        }

        let splitSdp = candidate.sdp.split(separator: " ")
        guard let ufragIndex = splitSdp.firstIndex(of: "ufrag") else {
            throw AttributeNotFoundError.UFragNotFound(
                description: "ufrag not found. Make sure the SDP answer is correct.")
        }
        let ufrag = String(splitSdp[ufragIndex + 1])

        let candidateDict =
            [
                "candidate": candidate.sdp,
                "sdpMLineIndex": candidate.sdpMLineIndex,
                "sdpMid": candidate.sdpMid ?? "",
                "usernameFragment": ufrag,
            ] as [String: Any]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: candidateDict, options: []) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "JSON serialization error"])
        }

        var components = URLComponents(string: connectOptions.serverUrl.absoluteString)
        components?.path = ""
        components?.path = patchEndpoint!
        let url = components?.url
        var request = URLRequest(url: url!)
        request.httpMethod = "PATCH"
        request.httpBody = jsonData
        request.setValue("application/trickle-ice-sdpfrag", forHTTPHeaderField: "Content-Type")

        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 204 else {
            throw SessionNetworkError.CandidateSendingError(
                description: "Candidate sending error - response was not successful.")
        }
    }

    public func peerConnection(_: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
        logger.debug("RTC signaling state changed: \(stateChanged.rawValue).")
    }

    public func peerConnection(_: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        DispatchQueue.main.async {
            if let track = stream.videoTracks.first {
                self.videoTrack = track
                self.delegate?.onTrackAdded(track: track)
            }
        }
        logger.debug("RTC media stream added: \(stream.description).")
    }

    public func peerConnection(_: RTCPeerConnection, didRemove stream: RTCMediaStream) {
        logger.debug("RTC media stream removed: \(stream.description).")
    }

    public func peerConnectionShouldNegotiate(_: RTCPeerConnection) {
        logger.debug("Peer connection negotiation needed.")
    }

    /**
      Reacts to changes in the ICE Connection state and logs a message depending on the current state.
     */
    public func peerConnection(_: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
        switch newState {
        case .checking:
            logger.debug("ICE is checking paths, this might take a moment.")
        case .connected:
            logger.debug("ICE has found a viable connection.")
        case .failed:
            logger.debug("No viable ICE paths found, consider a retry or using TURN.")
        case .disconnected:
            logger.debug("ICE connection was disconnected, attempting to reconnect or refresh.")
        case .new:
            logger.debug(
                "The ICE agent is gathering addresses or is waiting to be given remote candidates through calls")
        case .completed:
            logger.debug(
                "The ICE agent has finished gathering candidates, has checked all pairs against one another, and has found a connection for all components."
            )
        case .closed:
            logger.debug("The ICE agent for this RTCPeerConnection has shut down and is no longer handling requests.")
        default:
            break
        }
    }

    public func peerConnection(_: RTCPeerConnection, didChange newState: RTCIceGatheringState) {
        logger.debug("RTC ICE gathering state changed: \(newState.rawValue).")
    }

    /**
      Reacts to new candidate found and sends it to the WHIP/WHEP server.
     */
    public func peerConnection(_: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        if patchEndpoint != nil {
            Task { [weak self] in
                try await self?.sendCandidate(candidate: candidate)
            }
        } else {
            iceCandidates.append(candidate)
        }
    }

    public func peerConnection(_: RTCPeerConnection, didRemove _: [RTCIceCandidate]) {
        logger.debug("Removed candidate from candidates list.")
    }

    public func peerConnection(_: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
        logger.debug("RTC data channel opened: \(dataChannel.channelId)")
    }

    /**
      Reacts to changes in the Peer Connection state and logs a message depending on the current state
     */
    public func peerConnection(_: RTCPeerConnection, didChange stateChanged: RTCPeerConnectionState) {
        onConnectionStateChanged?(stateChanged)
        switch stateChanged {
        case .connected:
            logger.debug("Connection is fully connected")
        case .disconnected:
            logger.debug("One or more transports has disconnected unexpectedly")
        case .failed:
            logger.debug("One or more transports has encountered an error")
        case .closed:
            logger.debug("Connection has been closed")
        case .new:
            logger.debug("New connection")
        case .connecting:
            logger.debug("Connecting")
        default:
            logger.debug("Some other state: \(stateChanged.rawValue)")
        }
    }

    // MARK: - Codec Management

    /**
     Gets matched codecs from the preferred codec list that are available in the peer connection factory.
    
     - Parameter preferredCodecs: List of preferred codec names
     - Parameter mediaType: The media type (audio or video)
     - Parameter useReceiver: Whether to use receiver capabilities instead of sender capabilities
    
     - Returns: Array of matched codec capabilities
     */
    func getMatchedCodecs(preferredCodecs: [String], mediaType: String, useReceiver: Bool = false)
        -> [RTCRtpCodecCapability]
    {
        if preferredCodecs.isEmpty {
            return []
        }

        let capabilities =
            useReceiver
            ? ClientBase.peerConnectionFactory.rtpReceiverCapabilities(forKind: mediaType)
            : ClientBase.peerConnectionFactory.rtpSenderCapabilities(forKind: mediaType)
        let availableCodecs = capabilities.codecs

        return preferredCodecs.compactMap { preferredCodec in
            availableCodecs.first { codec in
                codec.name.caseInsensitiveCompare(preferredCodec) == .orderedSame
            }
        }
    }

    /**
     Sets codec preferences for a transceiver if matched codecs are available.
    
     - Parameter transceiver: The RTP transceiver to set preferences for
     - Parameter preferredCodecs: List of preferred codec names
     - Parameter mediaType: The media type (audio or video)
     - Parameter useReceiver: Whether to use receiver capabilities instead of sender capabilities
     */
    func setCodecPreferencesIfAvailable(
        transceiver: RTCRtpTransceiver?, preferredCodecs: [String], mediaType: String, useReceiver: Bool = false
    ) {
        let matchedCodecs = getMatchedCodecs(
            preferredCodecs: preferredCodecs, mediaType: mediaType, useReceiver: useReceiver
        )
        if !matchedCodecs.isEmpty {
            transceiver?.codecPreferences = matchedCodecs
        }
    }
}

File: packages/android-client/MobileWhepClient/src/main/java/com/mobilewhep/client/ClientBase.kt
package com.mobilewhep.client

import android.content.Context
import android.media.AudioAttributes
import android.util.Log
import com.mobilewhep.client.utils.PeerConnectionFactoryHelper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import org.webrtc.AudioTrack
import org.webrtc.DataChannel
import org.webrtc.IceCandidate
import org.webrtc.MediaStream
import org.webrtc.MediaStreamTrack
import org.webrtc.PeerConnection
import org.webrtc.RtpCapabilities
import org.webrtc.RtpReceiver
import org.webrtc.RtpTransceiver
import org.webrtc.VideoTrack
import org.webrtc.audio.AudioDeviceModule
import java.io.IOException
import java.net.ConnectException
import java.net.URI
import java.net.URL
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

internal const val CLIENT_TAG = "WHIP/WHEP Client"

interface ClientBaseListener {
  fun onTrackAdded(track: VideoTrack)
}

open class ClientBase(
  val appContext: Context,
  val stunServerUrl: String?
) : PeerConnection.Observer {

  protected val peerConnectionFactory = PeerConnectionFactoryHelper.getFactory(appContext)

  protected var peerConnection: PeerConnection? = null
  var connectOptions: ClientConnectOptions? = null

  val peerConnectionState: PeerConnection.PeerConnectionState?
    get() = peerConnection?.connectionState()


  protected var patchEndpoint: String? = null
  protected val iceCandidates = mutableListOf<IceCandidate>()

  protected val client = OkHttpClient()
  private val audioAttributes: AudioAttributes =
    AudioAttributes
      .Builder()
      .setUsage(AudioAttributes.USAGE_MEDIA)
      .setContentType(AudioAttributes.CONTENT_TYPE_MOVIE)
      .build()
  private val audioDeviceModule: AudioDeviceModule = createAudioDeviceModule(appContext, audioAttributes)

  private val coroutineScope: CoroutineScope =
    CoroutineScope(Dispatchers.Default)

  open var videoTrack: VideoTrack? = null
  open var audioTrack: AudioTrack? = null
  private var listeners = mutableListOf<ClientBaseListener>()
  var onTrackAdded: (() -> Unit)? = null
  var onConnectionStateChanged: ((PeerConnection.PeerConnectionState) -> Unit)? = null

  open fun setupPeerConnection() {
    val iceServers =
      listOf(
        PeerConnection.IceServer
          .builder(stunServerUrl ?: "stun:stun.l.google.com:19302")
          .createIceServer()
      )

    val config = PeerConnection.RTCConfiguration(iceServers)

    config.sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
    config.continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
    config.candidateNetworkPolicy = PeerConnection.CandidateNetworkPolicy.ALL
    config.tcpCandidatePolicy = PeerConnection.TcpCandidatePolicy.DISABLED

    try {
      peerConnection = peerConnectionFactory.createPeerConnection(config, this)!!

    } catch (e: NullPointerException) {
      throw SessionNetworkError.ConfigurationError("Failed to establish RTCPeerConnection. Check initial configuration")
    }
  }

  open suspend fun connect(connectOptions: ClientConnectOptions) {
    this.connectOptions = connectOptions
  }

  /**
   * Sends an SDP offer to the WHIP/WHEP server.
   *
   * @param sdpOffer - The offer to send to the server.
   *
   * @throws AttributeNotFoundError.ResponseNotFound if there is no response to the offer
   * @throws AttributeNotFoundError.LocationNotFound if the response does not contain the location parameter
   * @throws SessionNetworkError.ConnectionError if the  connection could not be established or the response code is incorrect,
   * for example due to server being down, wrong server URL or token.
   *
   * @return A SDP response.
   */
  suspend fun sendSdpOffer(sdpOffer: String) =
    suspendCoroutine { continuation ->
      if (connectOptions == null) {
        continuation.resumeWithException(
          SessionNetworkError.ConnectionError(
            "Cannot send the SDP Offer. Connection not setup. Remember to call connect first."
          )
        )
        return@suspendCoroutine
      }

      var requestBuilder: Request.Builder =
        Request
          .Builder()
          .url(connectOptions!!.serverUrl)
          .post(sdpOffer.toRequestBody())
          .header("Accept", "application/sdp")
          .header("Content-Type", "application/sdp")
      if (connectOptions!!.authToken != null) {
        requestBuilder =
          requestBuilder.header("Authorization", "Bearer " + connectOptions!!.authToken)
      }

      val request = requestBuilder.build()

      client.newCall(request).enqueue(
        object : Callback {
          override fun onFailure(
            call: Call,
            e: IOException
          ) {
            if (e is ConnectException) {
              continuation.resumeWithException(
                SessionNetworkError.ConnectionError(
                  "Network error. Check if the server is up and running and the token and the server url is correct."
                )
              )
            } else {
              Log.e(CLIENT_TAG, e.toString())
              continuation.resumeWithException(e)
              e.printStackTrace()
            }
          }

          override fun onResponse(
            call: Call,
            response: Response
          ) {
            response.use {
              patchEndpoint = response.headers["location"]

              if (patchEndpoint == null) {
                val exception =
                  AttributeNotFoundError.LocationNotFound(
                    "Location attribute not found. Check if the SDP answer contains location parameter."
                  )
                continuation.resumeWithException(exception)
                return
              }

              if (response.body == null) {
                val exception =
                  AttributeNotFoundError.ResponseNotFound("Response to SDP offer not found. Check if the network request was successful.")
                continuation.resumeWithException(exception)
                return
              }

              continuation.resume(response.body!!.string())
            }
          }
        }
      )
    }

  /**
   * Sends an ICE candidate to WHIP/WHEP server in order to provide a streaming device.
   *
   * @param candidate - Represents a single ICE candidate.
   *
   * @throws AttributeNotFoundError.PatchEndpointNotFound if the patch endpoint has not been properly set up
   * @throws AttributeNotFoundError.UFragNotFound if the SDP of the candidate does not contain the ufrag
   * @throws SessionNetworkError.CandidateSendingError` if the candidate could not be sent
   */
  suspend fun sendCandidate(candidate: IceCandidate) =
    suspendCoroutine { continuation ->
      if (connectOptions == null) {
        continuation.resumeWithException(
          SessionNetworkError.ConnectionError(
            "Cannot send the send ICE Candidate. Connection not setup. Remember to call connect first."
          )
        )
        return@suspendCoroutine
      }

      if (patchEndpoint == null) {
        continuation.resumeWithException(
          AttributeNotFoundError.PatchEndpointNotFound("Patch endpoint not found. Make sure the SDP answer is correct.")
        )
        return@suspendCoroutine
      }

      val splitSdp = candidate.sdp.split(" ")
      val ufrag = splitSdp[splitSdp.indexOf("ufrag") + 1]

      if (ufrag == null) {
        continuation.resumeWithException(AttributeNotFoundError.UFragNotFound("ufrag not found. Make sure the SDP answer is correct."))
        return@suspendCoroutine
      }

      val jsonObject = JSONObject()

      jsonObject.put("candidate", candidate.sdp)
      jsonObject.put("sdpMLineIndex", candidate.sdpMLineIndex)
      jsonObject.put("sdpMid", candidate.sdpMid)
      // TODO: is ufrag necessary or is it just elixir webrtc thing?
      jsonObject.put("usernameFragment", ufrag)

      val serverUrl = URL(connectOptions!!.serverUrl)
      val requestURL =
        URI(serverUrl.protocol, null, serverUrl.host, serverUrl.port, patchEndpoint, null, null)

      val request =
        Request
          .Builder()
          .url(requestURL.toURL())
          .patch(jsonObject.toString().toRequestBody())
          .header("Content-Type", "application/trickle-ice-sdpfrag")
          .build()

      client.newCall(request).enqueue(
        object : Callback {
          override fun onFailure(
            call: Call,
            e: IOException
          ) {
            continuation.resumeWithException(e)
            e.printStackTrace()
          }

          override fun onResponse(
            call: Call,
            response: Response
          ) {
            response.use {
              if (!it.isSuccessful) {
                continuation.resumeWithException(
                  SessionNetworkError.CandidateSendingError("Candidate sending error - response was not successful.")
                )
                return
              }
              continuation.resume(Unit)
            }
          }
        }
      )
    }

  override fun onSignalingChange(p0: PeerConnection.SignalingState?) {
    Log.d(CLIENT_TAG, "RTC signaling state changed:: ${p0?.name}")
  }

  protected fun getMatchedCodecs(
    preferredCodecs: List<String>,
    mediaType: MediaStreamTrack.MediaType,
    useReceiver: Boolean = false
  ): List<RtpCapabilities.CodecCapability> {
    if (preferredCodecs.isEmpty()) {
      return emptyList()
    }

    val availableCodecs =
      if (useReceiver) {
        peerConnectionFactory.getRtpReceiverCapabilities(mediaType).codecs
      } else {
        peerConnectionFactory.getRtpSenderCapabilities(mediaType).codecs
      }
    return preferredCodecs.mapNotNull { preferredCodec ->
      availableCodecs.firstOrNull { it.name.equals(preferredCodec, ignoreCase = true) }
    }
  }

  protected fun setCodecPreferencesIfAvailable(
    transceiver: RtpTransceiver?,
    preferredCodecs: List<String>,
    mediaType: MediaStreamTrack.MediaType,
    useReceiver: Boolean = false
  ) {
    val matchedCodecs = getMatchedCodecs(preferredCodecs, mediaType, useReceiver)
    if (matchedCodecs.isNotEmpty()) {
      transceiver?.setCodecPreferences(matchedCodecs)
    }
  }

  /**
   Reacts to changes in the ICE Connection state and logs a message depending on the current state.
   */
  override fun onIceConnectionChange(p0: PeerConnection.IceConnectionState?) {
    when (p0) {
      PeerConnection.IceConnectionState.NEW ->
        Log.d(
          CLIENT_TAG,
          "The ICE agent is gathering addresses or is waiting to be given remote candidates through calls"
        )

      PeerConnection.IceConnectionState.CHECKING ->
        Log.d(
          CLIENT_TAG,
          "ICE is checking paths, this might take a moment."
        )

      PeerConnection.IceConnectionState.CONNECTED ->
        Log.d(
          CLIENT_TAG,
          "ICE has found a viable connection."
        )

      PeerConnection.IceConnectionState.COMPLETED ->
        Log.d(
          CLIENT_TAG,
          "The ICE agent has finished gathering candidates, has checked all pairs against one another, " +
            "and has found a connection for all components."
        )

      PeerConnection.IceConnectionState.FAILED ->
        Log.d(
          CLIENT_TAG,
          "No viable ICE paths found, consider a retry or using TURN."
        )

      PeerConnection.IceConnectionState.DISCONNECTED ->
        Log.d(
          CLIENT_TAG,
          "ICE connection was disconnected, attempting to reconnect or refresh."
        )

      PeerConnection.IceConnectionState.CLOSED ->
        Log.d(
          CLIENT_TAG,
          "The ICE agent for this RTCPeerConnection has shut down and is no longer handling requests."
        )

      null -> Log.d(CLIENT_TAG, "The Peer Connection state is null.")
    }
  }

  override fun onIceConnectionReceivingChange(p0: Boolean) {
    Log.d(CLIENT_TAG, "onIceConnectionReceivingChange: $p0")
  }

  override fun onIceGatheringChange(p0: PeerConnection.IceGatheringState?) {
    Log.d(CLIENT_TAG, "RTC ICE gathering state changed: ${p0?.name}")
  }

  /**
   Reacts to new candidate found and sends it to the WHIP/WHEP server.
   */
  override fun onIceCandidate(candidate: IceCandidate) {
    if (patchEndpoint == null) {
      iceCandidates.add(candidate)
    } else {
      coroutineScope.launch {
        sendCandidate(candidate)
      }
    }
  }

  override fun onIceCandidatesRemoved(p0: Array<out IceCandidate>?) {
    Log.d(CLIENT_TAG, "Removed candidate from candidates list.")
  }

  override fun onAddStream(p0: MediaStream?) {
    Log.d(CLIENT_TAG, "RTC media stream added: ${p0?.id}")
  }

  override fun onRemoveStream(p0: MediaStream?) {
    Log.d(CLIENT_TAG, "RTC media stream removed: ${p0?.id}")
  }

  override fun onDataChannel(p0: DataChannel?) {
    Log.d(CLIENT_TAG, "RTC data channel opened: ${p0?.id()}")
  }

  override fun onRenegotiationNeeded() {
    Log.d(CLIENT_TAG, "Peer connection negotiation needed.")
  }

  /**
   Reacts to changes in the Peer Connection state and logs a message depending on the current state
   */
  override fun onConnectionChange(newState: PeerConnection.PeerConnectionState) {
    onConnectionStateChanged?.invoke(newState)
    when (newState) {
      PeerConnection.PeerConnectionState.NEW -> Log.d(CLIENT_TAG, "New connection")
      PeerConnection.PeerConnectionState.CONNECTING -> Log.d(CLIENT_TAG, "Connecting")
      PeerConnection.PeerConnectionState.CONNECTED -> Log.d(CLIENT_TAG, "Connection is fully connected")
      PeerConnection.PeerConnectionState.DISCONNECTED ->
        Log.d(
          CLIENT_TAG,
          "One or more transports has disconnected unexpectedly"
        )

      PeerConnection.PeerConnectionState.FAILED ->
        Log.d(
          CLIENT_TAG,
          "One or more transports has encountered an error"
        )

      PeerConnection.PeerConnectionState.CLOSED -> Log.d(CLIENT_TAG, "Connection has been closed")
    }
  }

  override fun onAddTrack(
    receiver: RtpReceiver?,
    mediaStreams: Array<out MediaStream>?
  ) {
    coroutineScope.launch(Dispatchers.Main) {
      val videoTrack = receiver?.track() as? VideoTrack?
      this@ClientBase.videoTrack = videoTrack
      listeners.forEach { listener -> videoTrack?.let { listener.onTrackAdded(it) } }
    }
    onTrackAdded?.let { it() }
  }

  fun addTrackListener(listener: ClientBaseListener) {
    listeners.add(listener)
    videoTrack?.let { listener.onTrackAdded(it) }
  }
}
